File: server/src/com/broll/gainea/server/core/cards/Card.java
package com.broll.gainea.server.core.cards;

import com.broll.gainea.net.NT_Card;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.actions.ActionHandlers;
import com.broll.gainea.server.core.actions.required.PlaceUnitAction;
import com.broll.gainea.server.core.actions.required.SelectChoiceAction;
import com.broll.gainea.server.core.player.Player;

public abstract class Card {

    protected GameContainer game;
    protected Player owner;
    private int id;
    private int picture;
    protected ActionHandlers actions;
    protected PlaceUnitAction placeUnitHandler;
    protected SelectChoiceAction selectHandler;

    private String title, text;
    private float drawChance = 1;

    public Card(int picture, String title, String text) {
        this.text = text;
        this.picture = picture;
        this.title = title;
    }

    protected void setDrawChance(float drawChance) {
        this.drawChance = drawChance;
    }

    public float getDrawChance() {
        return drawChance;
    }

    public void init(GameContainer game, Player owner, int id) {
        this.game = game;
        this.owner = owner;
        this.id = id;
    }

    public abstract boolean isPlayable();

    public void play(ActionHandlers actionHandlers) {
        this.actions = actionHandlers;
        placeUnitHandler = actions.getHandler(PlaceUnitAction.class);
        selectHandler = actions.getHandler(SelectChoiceAction.class);
        play();
    }

    protected abstract void play();

    public NT_Card nt() {
        NT_Card card = new NT_Card();
        card.id = (short) id;
        card.text = text;
        card.picture = (short) picture;
        card.title = title;
        return card;
    }

    public String getTitle() {
        return title;
    }

    public int getId() {
        return id;
    }

    public int getPicture() {
        return picture;
    }

    public String getText() {
        return text;
    }

    @Override
    public String toString() {
        return super.toString()+"{" +
                "owner=" + owner +
                ", title='" + title + '\'' +
                '}';
    }
}

File: server/src/main/com/broll/gainea/server/core/cards/Card.kt
package com.broll.gainea.server.core.cards

import com.broll.gainea.net.NT_Card
import com.broll.gainea.net.NT_Event_PlayedCard
import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.actions.ActionHandlers
import com.broll.gainea.server.core.actions.optional.CardAction
import com.broll.gainea.server.core.actions.required.PlaceUnitAction
import com.broll.gainea.server.core.actions.required.SelectChoiceAction
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral
import com.broll.gainea.server.core.utils.ProcessingUtils
import com.broll.gainea.server.core.utils.sendUpdate

abstract class Card(
    val picture: Int,
    val effectType: EffectType,
    val title: String,
    val text: String,
) {
    var id = 0
        private set
    protected lateinit var game: Game
    lateinit var owner: Player
    protected lateinit var actions: ActionHandlers
    protected lateinit var placeUnitHandler: PlaceUnitAction
    protected lateinit var selectHandler: SelectChoiceAction
    var drawChance = 1f
        protected set

    fun init(game: Game, owner: Player) {
        this.game = game
        this.owner = owner
        this.id = game.newObjectId()
    }

    open fun validFor(game: Game): Boolean = true

    abstract val isPlayable: Boolean
    fun play(actionHandlers: ActionHandlers) {
        actions = actionHandlers
        placeUnitHandler = actions.getHandler(PlaceUnitAction::class.java)
        selectHandler = actions.getHandler(SelectChoiceAction::class.java)
        play()
    }

    protected abstract fun play()
    open fun nt(): NT_Card {
        val card = NT_Card()
        card.id = id.toShort()
        card.text = text
        card.picture = picture.toShort()
        card.title = title
        return card
    }

    override fun toString(): String {
        return super.toString() + "{" +
                "owner=" + owner +
                ", title='" + title + '\'' +
                '}'
    }
}

fun Game.playCard(card: Card) {
    val nt = NT_Event_PlayedCard()
    if (card.owner.isNeutral()) {
        nt.player = -1
    } else {
        nt.player = card.owner.serverPlayer.id
    }
    nt.card = card.nt()
    if (card is DirectlyPlayedCard) {
        nt.sound = "glitter.ogg"
    } else {
        nt.sound = "dwooom.ogg"
    }
    sendUpdate(nt)
    card.owner.cardHandler.cards.remove(card)
    ProcessingUtils.pause(CardAction.PLAY_CARD_DELAY)
    card.play(reactionHandler.actionHandlers)
    updateReceiver.playedCard(card)
}
